File: Parcial2/civitasruby/lib/civitas/jugador_especulador.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas

  class Jugador_especulador < Jugador
    
    @@max_casas = @@factor_especulador * @@casas_max
    @@max_hoteles = @@factor_especulador * @@hoteles_max
    @@factor_pago_impuesto = 0.5
    
    def initialize(fianza)
      @fianza = fianza
    end
    
    def self.nuevo_especulador(otro,fianza)
      jugador = Jugador_especulador.new(fianza)
      jugador.copia(otro)
      jugador.actualizar_propiedades(otro)
      jugador
    end
    
    
    def actualizar_propiedades(original)
      original.get_propiedades().each do |propiedad|
        propiedad.actualiza_propietario_por_conversion(self)
      end
    end
    
    def to_s
      info = super
      info += "\n ---  JUGADOR ESPECULADOR: PAGA LA MITAD DE IMPUESTOS Y PUEDE PAGAR FIANZA EN LA CÁRCEL --- \n";
      info += "La fianza que paga es de #{@fianza}";
      return info;
    end
    
    def get_hoteles_max
      return @@max_hoteles
    end
    
    def get_casas_max
      return @@max_casas
    end
    
    def paga_impuesto(cantidad)
      paga(cantidad*@@factor_pago_impuesto)
    end
    
    def debe_ser_encarcelado
      carcel = super
      if (carcel)
        if (puedo_gastar(@fianza))
          paga(@fianza)
          carcel = false
        end
      end
      return carcel
    end
    
    
    
    
    
    
    
    
    
  end
end

File: Parcial2/Civitas/src/civitas/JugadorEspeculador.java
package civitas;

/**
 *
 * @author Salva
 */
public class JugadorEspeculador extends Jugador {

    private int maxCasas = FactorEspeculador * CasasMax;
    private int maxHoteles = FactorEspeculador * HotelesMax;

    private float fianza;
    private final float factorPagoImpuesto = 0.5f;

    JugadorEspeculador(Jugador otro, float fianza) {
        super(otro);
        this.fianza = fianza;
        actualizarPropiedades(otro);
    }

    //la dejamos o la quitamos?
    private void actualizarPropiedades(Jugador original) {
        original.getPropiedades().forEach((propiedad) -> {      // <- sugerido por NetBeans. Originalmente for (TituloPropiedad propiedad : original.getPropiedades()
            propiedad.actualizaPropietarioPorConversion(this);
        });
    }

    @Override
    public String toString() {
        String info = super.toString();
        info += "\n ---  Es jugador especulador. Paga menos impuestos y puede evitar la cárcel --- \n";
        info += "La fianza que paga es de " + fianza;
        return info;
    }

    @Override
    protected int getCasasMax() {
        return maxCasas;
    }

    @Override
    protected int getHotelesMax() {
        return maxHoteles;
    }

    @Override
    public boolean pagaImpuesto(float cantidad) {
        return super.pagaImpuesto(cantidad * factorPagoImpuesto);
    }

    @Override
    protected boolean debeSerEncarcelado() {
        boolean carcel = super.debeSerEncarcelado();
        if (carcel){
            if (puedoGastar(fianza)){
                paga(fianza);
                carcel = false;
            }
        }
        return carcel;
    }

}
